### ISSUE TITLE
Build fails with latest master branch of Emacs

### ISSUE THREAD

**Issue by yasuhirokimura** [K-Arakawa/emacs-ime-patch#2]
Emacs: 6094f3a16b of master branch
IME patch: emacs-master-ime.patch of a9fdf4e (latest)

Build on MSYS2 MINGW64 environment fails as following.

```
make[1]: Leaving directory '/c/Users/yasu/Work/Emacs/master/emacs/lib-src'
make -C src VCSWITNESS='$(srcdir)/../.git/logs/HEAD' BIN_DESTDIR=''/c/Emacs-master/bin/'' \
         ELN_DESTDIR='/c/Emacs-master/lib/emacs/29.0.50/' all
make[1]: Entering directory '/c/Users/yasu/Work/Emacs/master/emacs/src'
  GEN      globals.h
  CC       firstfile.o
  CC       dispnew.o
  CC       frame.o
frame.c: In function 'do_switch_frame':
frame.c:1463:15: error: incompatible types when initializing type 'int' using type 'specpdl_ref'
 1463 |   int count = SPECPDL_INDEX ();
      |               ^~~~~~~~~~~~~
frame.c:1467:14: error: incompatible type for argument 1 of 'unbind_to'
 1467 |   unbind_to (count, Qnil);
      |              ^~~~~
      |              |
      |              int
In file included from frame.c:29:
lisp.h:4452:31: note: expected 'specpdl_ref' but argument is of type 'int'
 4452 | extern Lisp_Object unbind_to (specpdl_ref, Lisp_Object);
      |                               ^~~~~~~~~~~
frame.c:1468:1: warning: control reaches end of non-void function [-Wreturn-type]
 1468 | }
      | ^
make[1]: *** [Makefile:414: frame.o] Error 1
make[1]: Leaving directory '/c/Users/yasu/Work/Emacs/master/emacs/src'
make: *** [Makefile:469: src] Error 2
```

**Comment by yasuhirokimura** [comment]
According to  the result of `git bisect` with `master` branch of Emacs repository, build fails with following commit or later.

```
commit 297520852f
Author: Mattias Engdegård <mattiase@acm.org>
Date:   Thu Feb 3 00:51:43 2022

    Turn specpdl_ref into a type of its own

    Wrap the specpdl_ref contents in a struct, which gives us some very
    useful type checking to make sure that all code that assumed integral
    specpdl refs has been converted.

    We only do this on 64-bit platforms, since those tend to have modern
    ABIs where small structs are optimised as scalars.  In other words,
    this change should not affect the compiled code.

    * src/lisp.h (specpdl_ref): Now a struct on 64-bit platforms.
    (wrap_specpdl_ref, unwrap_specpdl_ref): New.
    (specpdl_count_to_ref, specpdl_ref_to_count)
    (specpdl_ref_eq, specpdl_ref_lt)
    (specpdl_ref_valid_p, make_invalid_specpdl_ref): Use new wrappers.

diff --git a/src/lisp.h b/src/lisp.h
index 3ff9dccc4e..9b82307f67 100644
--- a/src/lisp.h
+++ b/src/lisp.h
@@ -3350,45 +3350,79 @@ #define DEFVAR_KBOARD(lname, vname, doc)                        \
     } bt;
   };

+/* We use 64-bit platforms as a proxy for ones with ABIs that treat
+   small structs efficiently.  */
+#if SIZE_MAX > 0xffffffff
+#define WRAP_SPECPDL_REF 1
+#endif
+
 /* Abstract reference to to a specpdl entry.  */
+#ifdef WRAP_SPECPDL_REF
+/* Use a proper type for specpdl_ref if it does not make the code slower,
+   since the type checking is quite useful.  */
+typedef struct { ptrdiff_t n; } specpdl_ref;
+#else
 typedef ptrdiff_t specpdl_ref;
+#endif

+/* Internal use only.  */
 INLINE specpdl_ref
-specpdl_count_to_ref (ptrdiff_t count)
+wrap_specpdl_ref (ptrdiff_t count)
 {
+#ifdef WRAP_SPECPDL_REF
+  return (specpdl_ref) {.n = count};
+#else
   return count;
+#endif
 }
+/* Internal use only.  */
 INLINE ptrdiff_t
-specpdl_ref_to_count (specpdl_ref ref)
+unwrap_specpdl_ref (specpdl_ref ref)
 {
+#ifdef WRAP_SPECPDL_REF
+  return ref.n;
+#else
   return ref;
+#endif
+}
+
+INLINE specpdl_ref
+specpdl_count_to_ref (ptrdiff_t count)
+{
+  return wrap_specpdl_ref (count);
+}
+
+INLINE ptrdiff_t
+specpdl_ref_to_count (specpdl_ref ref)
+{
+  return unwrap_specpdl_ref (ref);
 }

 /* Whether two `specpdl_ref' refer to the same entry.  */
 INLINE bool
 specpdl_ref_eq (specpdl_ref a, specpdl_ref b)
 {
-  return a == b;
+  return unwrap_specpdl_ref (a) == unwrap_specpdl_ref (b);
 }


 /* Whether `a' refers to an earlier entry than `b'.  */
 INLINE bool
 specpdl_ref_lt (specpdl_ref a, specpdl_ref b)
 {
-  return a < b;
+  return unwrap_specpdl_ref (a) < unwrap_specpdl_ref (b);
 }

 INLINE bool
 specpdl_ref_valid_p (specpdl_ref ref)
 {
-  return specpdl_ref_to_count (ref) >= 0;
+  return unwrap_specpdl_ref (ref) >= 0;
 }

 INLINE specpdl_ref
 make_invalid_specpdl_ref (void)
 {
-  return specpdl_count_to_ref (-1);
+  return wrap_specpdl_ref (-1);
 }

 /* Return a reference that is `delta' steps more recent than `ref'.
```

**Comment by yasuhirokimura** [comment]
Build error is fixed with 4b205cd.
